src/lex/cpp.rs

```diff
@@ -1152,6 +1158,25 @@ impl<'a> PreProcessor<'a> {
         }
         Ok(ret)
     }
+
+    fn update_builtin_definitions(&mut self) {
+        use Definition::Object;
+        let int_def = |i: i32| Object(vec![LiteralToken::Int(RcStr::from(i.to_string())).into()]);
+        let str_def = |s: &str| {
+            let rcstr = RcStr::from(format!("\"{}\"", s));
+            Object(vec![LiteralToken::Str(vec![rcstr]).into()])
+        };
+        self.definitions.extend(map! {
+            "__LINE__".into() => int_def((self.line() + 1) as i32),
+            "__FILE__".into() => str_def(self.file_processor.path().to_str().unwrap()),  // Fails if non-Unicode
+            "__DATE__".into() => {
+                str_def(&time::OffsetDateTime::now_local().format("%b %_d %Y"))
+            },
+            "__TIME__".into() => {
+                str_def(&time::OffsetDateTime::now_local().format("%H:%M:%S"))
+            },
```

Comment:
The standard says DATE and TIME should remain the same throughout compilation: http://port70.net/~nsz/c/c11/n1570.html#6.10.8. So this should be moved to `Preprocessor::new` I think.

Reply:
Ah, didn't see that. I'll change it. This also makes testing `__DATE__` and `__TIME__` much easier because I can mock the preprocessor to have certain times to ensure the formatting is correct.